bootstrap/bootstrap.go

```diff
@@ -1398,31 +1398,48 @@ func (b *Bootstrap) defaultCheckoutPhase(ctx context.Context) error {
 			gitVersionOutput, _ := b.shell.RunAndCapture(ctx, "git", "--version")
 			b.shell.Warningf("Failed to recursively sync git submodules. This is most likely because you have an older version of git installed (" + gitVersionOutput + ") and you need version 1.8.1 and above. If you're using submodules, it's highly recommended you upgrade if you can.")
 		}
-
+		args := []string{}
+		for _, config := range b.GitSubmoduleCloneConfig {
+			args = append(args, "-c", config)
+		}
 		// Checking for submodule repositories
 		submoduleRepos, err := gitEnumerateSubmoduleURLs(ctx, b.shell)
 		if err != nil {
 			b.shell.Warningf("Failed to enumerate git submodules: %v", err)
 		} else {
-			for _, repository := range submoduleRepos {
+			for idx, repository := range submoduleRepos {
 				// submodules might need their fingerprints verified too
 				if b.SSHKeyscan {
 					addRepositoryHostToSSHKnownHosts(ctx, b.shell, repository)
 				}
-			}
-		}
-
-		args := []string{}
-		for _, config := range b.GitSubmoduleCloneConfig {
-			args = append(args, "-c", config)
-		}
-		args = append(args, "submodule", "update", "--init", "--recursive", "--force")
-		if err := b.shell.Run(ctx, "git", args...); err != nil {
-			return err
-		}
+				if mirrorDir != "" {
+					name := fmt.Sprintf("submodule-%d", idx+1)
+					if err := b.shell.Run(ctx, "git", "--git-dir", mirrorDir, "remote", "add", name, repository); err != nil {
+						// Per https://git-scm.com/docs/git-remote#_exit_status: When the remote already exists, the exit status is 3
+						// That shouldn't be a fatal error in this case, so ignore it.
+						if err.Error() != "exit status 3" {
+							return err
+						}
+					}
```

Comment:
Hmmm, I looked at the `shell.GetExitCode(err)` method, and it does not inspire a log of confidence that it will work on Windows. But I've made #1938 to make it more cross-platform.

It should be fine to merge this PR with my suggestion after that's merged, but I think it would be safest to rebase.

Reply:
Thanks for the quick feedback @triarius, your suggestion is much cleaner 👍🏻 

Would you prefer that I used the magic "Commit suggestion" button or apply your suggestion to my local branch, rebase and force push?

FWIW when I test it on Windows it works as expected with or without your changes from #1938 but I don't have any objections to waiting on that getting merged.

Reply:
> Would you prefer that I used the magic "Commit suggestion" button or apply your suggestion to my local branch, rebase and force push?

That's fine by me. I will run CI on this again before merging.

> FWIW when I test it on Windows it works as expected with or without your changes from https://github.com/buildkite/agent/pull/1938 but I don't have any objections to waiting on that getting merged.

Oh nice, #1938 is merged now in any case.